### DISCUSSION TITLE
Securing use of third-party GitHub actions

### DISCUSSION THREAD

**Discussion by solarmosaic-kflorence** [community/community#26310] [Actions]
<p>In light of <a href="https://github.blog/2021-11-15-githubs-commitment-to-npm-ecosystem-security/" class="inline-onebox" rel="noopener nofollow ugc">GitHub's commitment to npm ecosystem security | The GitHub Blog</a> – I am wondering how GitHub plans to ensure the GitHub actions ecosystem also remains secure in the event of third-party action repository takeovers. As suggested in <a href="https://docs.github.com/en/actions/security-guides/security-hardening-for-github-actions#using-third-party-actions" class="inline-onebox" rel="noopener nofollow ugc">Security hardening for GitHub Actions - GitHub Docs</a> – the safest way at the moment is to pin any third-party actions to a commit SHA, but unfortunately most documentation on GitHub and in third-party actions suggest to pin to a major version (e.g. <code>@v1</code>), which means it will automatically pull in subsequent tags for that major version.</p>
<p>This is especially troubling for third-party actions which are invoked in an authenticated context. They could not only access the <code>GITHUB_TOKEN</code>, but it could also compromise cloud-accounts such as AWS if precautions are not taken.</p>
<p>These are problems that have already been solved in other ecosystems, through things like caching layers (to guarantee the resource does not change), multi-factor verification when publishing new versions of a package, and ensuring that previously published packages cannot be modified. At the very least, it seems that GitHub should promote security best practices in all places in their documentation, instead of expecting someone to find and read the security section. Even better if this could be enforced through a process or convention.</p>
<p>PS: this is even more of a nightmare when using an action that itself references other actions. The only way to get around that is to fork the action and maintain it yourself, or copy the action into your repository.</p>

**Comment by ViacheslavKudinov** [comment]
<p>This could be utterly beneficial for Enterprise organizations.<br>
At the moment all the time have to do these assessments by ourselves for each new version and it is still tricky process.</p>

**Comment by solarmosaic-kflorence** [comment]
<p>At the very least, providing some kind of manifest/lock file that could be defined in the <code>.github</code> folder, which would apply to all actions (even those referenced within actions), would be very useful. Currently we have to maintain commit SHAs across hundreds of workflow files in our monorepo and this would greatly improve that experience.</p>

**Comment by irongut** [comment]
<p>As the author of a couple of Actions I would suggest pinning to a full tagged version number like v1.1.0, pinning to a commit SHA is also fine if it is a tagged version. Nothing else is guaranteed to work and for my actions probably won’t due to use of Docker and GHPR.</p>
<p>I would like MFA but it must work easily in a release workflow.</p>

**Comment by pauleustice** [comment]
<p>I feel your pain! I don’t know if you’ve done this already, but you could create your own manifest file and read it into each workflow; I haven’t tried this but I guess it would work.</p>
<p>Each workflow would need to:</p>
<ul>
<li>Use <code>actions/checkout</code> to clone the repo (incl manifest)</li>
<li>Read the manifest (maybe via <code>run</code> or github-script)</li>
<li>Store the manifest content as an env</li>
<li>Reference the relevant value in the <code>uses</code> value of each action</li>
</ul>

**Comment by solarmosaic-kflorence** [comment]
<p>Yes, that would work. We have done something similar for other things (for example using <code>.nvmrc</code> to read the Node version). I guess it is the best workaround for now, even though it adds a bit of complexity to our workflows.</p>
<p>It wouldn’t solve the versions used in actions we consume, though. Only the versions we reference in our own actions.</p>

**Comment by solarmosaic-kflorence** [comment, marked as answer]
<p>Moved this discussion here: <a href="https://github.com/github-community/community/discussions/18103" class="inline-onebox" rel="noopener nofollow ugc">Securing use of third-party GitHub actions · Discussion #18103 · github-community/community · GitHub</a></p>

**Comment by Arcitec** [comment]
> unfortunately most documentation on GitHub and in third-party actions suggest to pin to a major version (e.g. `@v1`), which means it will automatically pull in subsequent tags for that major version.

That's not true. `@v1` literally means the tag named `v1`. It's up to the upstream repo whether they will point that to v1.0.0 or the latest tag for that major version (ie v1.5.3). I can confirm that Docker and GitHub's action repos re-tag it to point at the latest revision. But it's absolutely not a rule. It's something those repos have to implement themselves.
